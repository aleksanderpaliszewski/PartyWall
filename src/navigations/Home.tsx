import React from 'react';
import {Scenes} from '../api/enums';
import {
  createStackNavigator,
  StackNavigationProp,
} from '@react-navigation/stack';
import {RouteProp} from '@react-navigation/native';
import Home from '../scenes/Home';

export type ScreenName = Scenes.Home;

export type StackParamList = {
  [Scenes.Home]: undefined;
};

export interface HomeStackScreenProps<T extends ScreenName> {
  navigation: StackNavigationProp<StackParamList, T>;
  route: RouteProp<StackParamList, T>;
}

const {Navigator, Screen} = createStackNavigator<StackParamList>();

const HomeStackScreen = () => {
  return (
    <Navigator
      screenOptions={{
        headerShown: false,
      }}>
      <Screen name={Scenes.Home} component={Home} />
    </Navigator>
  );
};

export default HomeStackScreen;
