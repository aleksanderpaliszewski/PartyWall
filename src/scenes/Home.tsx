import React, {FC, useContext, useState, useEffect} from 'react';
import _ from 'lodash';

import ProductList from '../components/ProductList';
import UserContext from '../utils/userContext';
import {HomeStackScreenProps} from '../navigations/Home';
import {useFetch} from '../hooks/useFetch';
import {Scenes} from '../api/enums';
import {ProductType} from '../api/types';
import {Drink, Food} from '../api/interface';

export const PRODUCTS_PER_SCREEN = 10;

const Home: FC<HomeStackScreenProps<Scenes.Home>> = () => {
  const {resetAuthData} = useContext(UserContext);
  const [products, setProducts] = useState<ProductType[]>([]);
  const [offset, setOffset] = useState(0);

  const {data, updateUrl, reload, isLoading} = useFetch<ProductType[]>(
    `/products?_start=${offset}&_limit=${PRODUCTS_PER_SCREEN}`,
    [],
  );

  useEffect(() => {
    if (!data) {
      return;
    } else if (offset > 0) {
      setProducts((products) => _.union(products, data));
    } else {
      setProducts(data);
    }
  }, [data, updateUrl, offset]);

  const fetchMoreProducts = () => {
    const newOffset = offset + 10;
    setOffset(newOffset);

    if (data?.length) {
      updateUrl(`/products?_start=${newOffset}&_limit=${PRODUCTS_PER_SCREEN}`);
    }
  };

  const isDrink = (product: Food | Drink): product is Drink =>
    !!(product as Drink).volume;

  return (
    <ProductList
      isLoading={isLoading}
      products={products}
      reload={reload}
      resetAuthData={resetAuthData}
      fetchMoreProducts={fetchMoreProducts}
      isDrink={isDrink}
    />
  );
};

export default Home;
