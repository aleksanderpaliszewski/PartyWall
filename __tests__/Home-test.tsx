import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import {shallow, configure} from 'enzyme';

import Home from '../src/scenes/Home';
import {getHomeProps} from './navigation';

configure({adapter: new Adapter()});
describe('Home', () => {
  it('render', () => {
    const mockProps = getHomeProps();

    const wrapper = shallow(<Home {...mockProps} />);
    expect(wrapper).toMatchSnapshot();
  });
});
