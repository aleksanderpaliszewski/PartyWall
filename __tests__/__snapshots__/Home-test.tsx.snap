// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home render 1`] = `
<ProductList
  fetchMoreProducts={[Function]}
  isDrink={[Function]}
  isLoading={false}
  products={Array []}
  reload={[Function]}
  resetAuthData={[Function]}
/>
`;
